File: interviewPractice/java/arrays/Sudoku2.java
import java.util.*;

public class Sudoku2 {
	boolean sudoku2(char[][] grid) {
		if(grid.length != 9 || grid[0].length != 9) {
			return false;
		}

		// check each row
		for(int i = 0; i < 9; i++) {
			Set<Character> rowValues = new HashSet<>();
			for(int j = 0; j < 9; j++) {
				if(grid[i][j] != '.') {
					if(rowValues.contains(grid[i][j])) {
						return false;
					}
					rowValues.add(grid[i][j]);
				}
			}
		}

		// check each column
		for(int i = 0; i < 9; i++) {
			Set<Character> columnValues = new HashSet<>();
			for(int j = 0; j < 9; j++) {
				if(grid[j][i] != '.') {
					if(columnValues.contains(grid[j][i])) {
						return false;
					}
					columnValues.add(grid[j][i]);
				}
			}
		}

		// check each subgrid
		for(int i = 0; i < 9; i += 3) {
			for(int j = 0; j < 9; j += 3) {
				// we are at the top left corner of the subgrid
				Set<Character> subgridValues = new HashSet<>();
				for(int k = i; k < i + 3; k++) {
					for(int m = j; m < j + 3; m++) {
						if(grid[k][m] != '.') {
							if(subgridValues.contains(grid[k][m])) {
								return false;
							}
							subgridValues.add(grid[k][m]);
						}
					}
				}
			}
		}

		return true;
	}

	public static void main(String[] args) {
		char[][] grid = {{'.', '.', '.', '.', '2', '.', '.', '9', '.'},
        {'.', '.', '.', '.', '6', '.', '.', '.', '.'},
        {'7', '1', '.', '.', '7', '5', '.', '.', '.'},
        {'.', '7', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '8', '3', '.', '.', '.'},
        {'.', '.', '8', '.', '.', '7', '.', '6', '.'},
        {'.', '.', '.', '.', '.', '2', '.', '.', '.'},
        {'.', '1', '.', '2', '.', '.', '.', '.', '.'},
        {'.', '2', '.', '.', '3', '.', '.', '.', '.'}};

    System.out.println(new Sudoku2().sudoku2(grid));
	}
}

File: interviewPractice/python/arrays/sudoku_2.py
def valid_values(values):
	numbers = list(filter((lambda value: value != '.'), values))
	return False if len(numbers) > len(set(numbers)) else True

def sudoku_2(grid):
	for row in grid:
		if not valid_values(row):
			return False

	for i in range(0, len(grid[0])):
		if not valid_values([row[i] for row in grid]):
			return False

	for i in range(0, 3):
		for j in range(0, 3):
			subgrid = []
			for m in range(0, 3):
				for n in range(0, 3):
					subgrid.append(grid[i * 3 + m][j * 3 + n])

			if not valid_values(subgrid):
				return False

	return True

if __name__ == '__main__':
	print(sudoku_2(
		[[".",".",".","1","4",".",".","2","."],
		 [".",".","6",".",".",".",".",".","."],
		 [".",".",".",".",".",".",".",".","."],
		 [".",".","1",".",".",".",".",".","."],
		 [".","6","7",".",".",".",".",".","9"],
		 [".",".",".",".",".",".","8","1","."],
		 [".","3",".",".",".",".",".",".","6"],
		 [".",".",".",".",".","7",".",".","."],
		 [".",".",".","5",".",".",".","7","."]]
		) == True)

	print(sudoku_2(
		[[".",".",".",".","2",".",".","9","."],
		 [".",".",".",".","6",".",".",".","."],
		 ["7","1",".",".","7","5",".",".","."],
		 [".","7",".",".",".",".",".",".","."],
		 [".",".",".",".","8","3",".",".","."],
		 [".",".","8",".",".","7",".","6","."],
		 [".",".",".",".",".","2",".",".","."],
		 [".","1",".","2",".",".",".",".","."],
		 [".","2",".",".","3",".",".",".","."]]
		 ) == False)
